### ISSUE TITLE
(Suggestion) Add a way for users to provide password file parsing logic

### ISSUE THREAD

**Issue by gandalf3** [browserpass/browserpass-native#98]
I have carefully curated my password store to be parsed by scripts I've written, which gives me enough flexibility to handle any edge cases which do eventually appear. While browserpass's generic rules do work for me most of the time, occasionally there are exceptions. For instance:

* I have one site which *requires* an empty username field. I haven't found a way to convince browserpass to use an empty string as the login.
* Many sites specifically ask for an email address as the login. I like to keep my password metadata normalized, so I have my own syntax and logic for determining when the email field should be used as the login.

Would it be reasonable to add configuration to browserpass which would permit a command to be called, either including the pass invocation or just some program which takes pass's output in stdin and prints the appropriate field to stdout?

As in something like:

```
#login: "pass show ${login_path} | parse_login.sh"
# or just
login: "parse_login.sh"
# I do have all my passwords on the first line, so I personally only want to override login for now, but other people might have more complex use cases perhaps
#password: "parse_password.sh"
```

Is this something which suits the mission of browserpass?

By the way. thanks for writing this excellent extension!

**Comment by erayd** [comment]
I can see where you're coming from. There are a few issues though:

1. We don't actually call `pass` at all. We just use `gpg` to decrypt, and return the entire decrypted contents to Browserpass.
2. We have logic that addresses the missing-username case, and uses the filename under some circumstances. This would conflict. See #100 and #225 for additional context, and why we don't use `email` as a login field.
3. We will eventually be *writing* (end encrypting) pass entries. If a filter program has modified them first, this would mean we'd end up writing modified data back into the original file, thus corrupting it.

Your case of sites that require a password but explicitly do *not* take a username is an important one. I've hit the odd one of these too. @maximbaz - how would you feel about us adding a per-entry `nullLogin` option, which would mean "do not attempt to fill a username for this entry"?

Regarding data normalisation - I can understand this desire. However, is there any reason why you cannot simply use the following formats to handle both sites that require an email login, and sites that require a username login, while still keeping a standard email field?

```
myPassword
login: myUsername
email: user@example.com
```
```
myPassword
login: user@example.com
email: user@example.com
```

If the `email` field is always present when the entry contains an email address, and the `login` field always means "use this thing as the username", it seems to me that this would address your issue.

Are there any other cases which you think would necessitate such a filter script?

If we decide not to add a filter script option (which to be frank is the most likely outcome, unless you can present a **really** compelling case), note that you can always write a wrapper for `gpg` and point Browserpass at it. We provide an option to set a custom `gpg` binary, so if you were so inclined you could certainly abuse that feature to achieve what you're wanting to do here.

**Comment by gandalf3** [comment]
Thanks for the detailed response! I see it's not a simple problem.

TL;DR: The gpg override (plus some kind of nullLogin flag) would be good enough for me, thanks!

----

I would be comfortable hijacking the gpg override option to do what I want (I hadn't thought of that possibility, thanks for pointing it out!). I think I can even handle writeback that way? I'm assuming something like this will work:
* When browserpass calls `gpg --decrypt -`; my wrapper can take gpg's output here make it browserpass-friendly.
* When browserpass calls `gpg --encrypt -r $gpg_id` it'll pass the modified browserpass-friendly text on stdin; my wrapper can decrypt the actual file again, merge the new fields from the browserpass-friendly text, and encrypt the result. Does that seem about right? If a `nulllogin` or similar feature were added, then I think this would cover my current use cases.

That said, the increased complexity of doing things as a gpg wrapper does perhaps increase the potential of user-provided bugs, and is subject to breakage if browserpass starts calling gpg differently in the future. I could also perhaps imagine a desire to have different parsing rules on different password stores (which I don't think a gpg-wrapper would have any way to detect, given the file is passed via stdin).

Perhaps it would be possible to ask the user for a script to insert data back into the file, in the event writing is to be enabled and a filter script is used? e.g.

```
login_read: "browserparse.sh --get --login"
login_write: "browserparse.sh --set --login" # if login_read is provided but login_write is not, writing is disabled
# password_read:
# password_write:
```

If that is too many options, what about having the user provided script return the start and end positions of the requested string, rather than the string itself? It might be a little unusual, and would only work if the strings are actually stored contiguously (my setup isn't quite this odd, but I could *maybe* see some temptation to, say, construct an email from username + filename). But it would be simpler in the sense that there would be fewer configuration options.

My argument for wrapper support is: I already have about 100 passwords formatted to work with my script, and while I could probably automate the addition of an login-field clone of any email fields, I'd much prefer to write a wrapper for a few reasons:

* Definitely no other problems caused by incompatibilities between my formatting and browserpass's interpretation will appear (for instance, some of my fields are quoted).
* No denormalization means I'll never accidentally update an email address one way and not another way, leading to different tools producing different logins.
* I'm lazy and writing my own browserpass compatibility layer seems to me less effort than updating most of my password store *and* adapting my existing tools wherever there's a conflict in behavior.

I understand that probably isn't enough to justify it, given such a wrapper is pretty much possible already. Nevertheless I thought I'd share my thoughts, as I expect I'm probably not the only pass user in this kind of situation.

**Comment by maximbaz** [comment]
Interesting thread 🙂 

Instead of hijacking `gpg` let me ask you a different question: why don't you re-implement the native host entirely? It is actually very small app, probably smaller than you are anticipating, we intentionally designed the architecture so that native host remains minimal and all the fancy logic is offloaded to the browser extension.

If you like Go, you could just take this repo as a baseline and adjust according to your own rules. If you want to use bash or any other language, that is certainly possible. It's almost what you want, take an input and write output, only instead of stdin/stdout it's a _liiiitle_ more more different, due to the protocol itself.

We actually have the API documented entirely: https://github.com/browserpass/browserpass-native/blob/master/PROTOCOL.md

And you don't even have to implement 100% of the features if you don't use them anyway! For example, you don't have to respect "settings" object if all the paths are hardcoded in your native host anyway.

And you don't have to use `gpg`, your implementation can call the `pass` for example.

Possibilities are endless 😉 

What do you think?

> @maximbaz - how would you feel about us adding a per-entry nullLogin option, which would mean "do not attempt to fill a username for this entry"?

Agree, although maybe with a more intuitive syntax that we wouldn't even need to document? Could we try to support empty value instead (as below), i.e. distinguish between an absent `login:` and a present one with an empty value?

```
mypassword
login:
```

**Comment by gandalf3** [comment]
That's even better! My main concern with hijacking gpg was the possibility of it breaking in strange ways as things update. Having an actual documented API (!), even if it's not really guaranteed to be static, definitely helps alleviate that concern. Thanks!

**Comment by erayd** [comment]
@maximbaz - allowing an empty `login:` seems like an ideal way to implement this. I agree with you. It means we'll need to handle it as an edge case (Browserpass currently treats empty or all-whitespace values as nonexistent), but that should be fairly straightforward. I'll do a PR for it.

@gandalf3 - while it's not guaranteed to be static, it *is* intended to be extremely backwards-compatible. We designed it so that we should only very rarely need to update the host app, and the intention is that when we do, upgrading it should not be required in order for Browserpass to work. Obviously any *new* functionality that requires changes to the native host won't work without the native host being updated, but failing to upgrade (or sticking with an older version of the spec) should not *break* anything. That was one of our biggest considerations when we designed the new v3 architecture for Browserpass. Host update requirements have been a massive headache in the past, and we wanted to solve that.

If you do decide to modify the native host, please ensure that it reports an accurate major & minor version number for the version of the spec it's using. Browserpass uses that in order to determine the capabilities of the native host, and this will ensure we don't ask your implementation of it to do anything it doesn't understand.

**Comment by maximbaz** [comment]
Empty login is implemented, seems like we can close this issue now 🙂
